fw/client.c

```diff
@@ -66,22 +67,87 @@ typedef struct {
 	char		user_agent[UA_CMP_LEN];
 } TfwClientEntry;
 
+typedef struct {
+	struct list_head	list;
+	unsigned long		key;
+} TfwClientLRU;
+
+static struct {
+	struct list_head	head;
+	struct list_head	freelist;
+	TfwClientLRU		*mem;
+} client_lru;
+
 static TDB *client_db;
 
+static void
+tfw_client_update_lru(unsigned long key)
+{
+	if (list_empty(&client_lru.freelist)) {
+		/* Free list is empty, get the last from LRU list. */
+		TfwClientLRU *last = list_last_entry(&client_lru.head,
+						     TfwClientLRU, list);
+
+		list_del_init(&last->list);
+		tdb_entry_remove(client_db, last->key, NULL, NULL, false);
+		last->key = key;
+		list_add(&last->list, &client_lru.head);
+	} else {
+		TfwClientLRU *new = list_first_entry(&client_lru.freelist,
+						     TfwClientLRU, list);
+
+		list_del_init(&new->list);
+		new->key = key;
+		list_add(&new->list, &client_lru.head);
+	}
+}
+
+static int
+tfw_client_init_lru(void)
+{
+	TfwClientLRU *block;
+	int i, order = get_order(sizeof(TfwClientLRU) * client_cfg.lru_size);
+
+	INIT_LIST_HEAD(&client_lru.freelist);
+	INIT_LIST_HEAD(&client_lru.head);
+
+	client_lru.mem = (TfwClientLRU *)__get_free_pages(GFP_ATOMIC, order);
+	if (!client_lru.mem) {
+		T_ERR("Can't allocate client LRU list.");
+		return -ENOMEM;
+	}
+
+	for (i = 0, block = client_lru.mem; i < client_cfg.lru_size; i++)
+		list_add(&block[i].list, &client_lru.freelist);
+
+	return 0;
+}
+
+static void
+tfw_client_free_lru(void)
+{
+	int order = get_order(sizeof(TfwClientLRU) * client_cfg.lru_size);
+
+	free_pages((unsigned long)client_lru.mem, order);
+}
+
 /**
  * Called when a client socket is closed.
  */
 void
 tfw_client_put(TfwClient *cli)
 {
 	TfwClientEntry *ent = (TfwClientEntry *)cli;
+	TdbFRec *rec = ((TdbFRec *)cli) - 1;
 
 	T_DBG2("put client %p, users=%d\n",
 	       cli, ent->users);
 
 	spin_lock(&ent->lock);
 
 	--ent->users;
+	if (ent->users > 0)
+		tdb_rec_put(client_db, rec);
```

Comment:
May be move it just after `if (likely(ent->users)) { ` on string 153. Why we check  `if (ent->users > 0)` two times?